### PR TITLE
Test: wait for network changes in TestNetworkDBNodeJoinLeaveIteration

### LINKED ISSUE #42698: Flaky test: TestNetworkDBNodeJoinLeaveIteration
Looks like https://github.com/moby/moby/pull/42625#issuecomment-878447879 didn't fully resolve the flakiness in this test;

```
 === FAIL: libnetwork/networkdb TestNetworkDBNodeJoinLeaveIteration (5.81s)
[2021-07-29T18:12:43.071Z]     networkdb_test.go:511: Network existence verification failed
[2021-07-29T18:12:43.071Z]     networkdb_test.go:513: The networkNodes list has to have be 2 instead of 1 - [8518394491d6]
```

### PR DESCRIPTION
**Pull request by zq-david-wang**
Signed-off-by: David Wang <00107082@163.com>

fix #42698
In network node change test, the expected behavior is focused on how many nodes left in networkDB,  besides timing issues, things would also go tricky for a leave-then-join sequence,  if the check (counting the nodes) happened before the first "leave" event, then the testcase actually miss its target and report PASS without verifying its final result; if the check happened after the 'leave' event, but before the 'join' event, the test would report FAIL unnecessary;

This code change would check both the db changes and the node count, it would report PASS only when networkdb has indeed changed and the node count is expected

### CODE REVIEW

**Comment by thaJeztah** [comment]
Thanks for contributing @zq-david-wang - and apologies for the long delay; I was trying to get more eyes on this fix, but it slipped off my radar. Let me try to get some eyes on this one.

I did a quick rebase of your PR to make CI run again, and I included your description from GitHub in the commit message (which can be useful when looking through git history).

**Comment by zq-david-wang** [comment]
@thaJeztah  I am not sure whether this PR is still necessary, after such a long period, any recent random failure for this test case? 
Any way, I make some adjustment  base on  @corhere comments.

It is ok for me to close this if the PR is not needed anymore, otherwise @corhere , could you help review the new code, thx~

**Comment by thaJeztah** [comment]
I think the test was still flaky (at least to some extend), and (IIRC) mostly waiting for the suggestions to be addressed; so thanks for updating it!

@corhere ptal

**Review by thaJeztah — APPROVED** [review]
LGTM

**Review by corhere — CHANGES_REQUESTED** [review]

File: libnetwork/networkdb/networkdb_test.go [hunk 1]
```diff
@@ -479,22 +480,47 @@ func TestNetworkDBCRUDMediumCluster(t *testing.T) {
 func TestNetworkDBNodeJoinLeaveIteration(t *testing.T) {
 	dbs := createNetworkDBInstances(t, 2, "node", DefaultConfig())
 
+	var (
+		dbIndex          int32
+		staleNetworkTime [2]serf.LamportTime
+		expectNodeCount  int
+		network          = "network1"
+	)
+	dbChangeWitness := func(t poll.LogT) poll.Result {
```
[the remark this answers is not in the document]
Comment (corhere) [reply libnetwork/networkdb/networkdb_test.go:489]: This can be addressed in a follow-up PR since it's all private to the one test

**Review by corhere — APPROVED** [review]

**Comment by thaJeztah** [comment]
Thanks! Let's get this one in 👍

**Pull request merged**